File: javascript/anagram/anagram.js
export class Anagram {
  constructor(subject) {
    this.subject = subject;
  }

  matches(candidates) {
    return candidates.filter(candidate => {
      return (
        candidate.toLowerCase() !== this.subject.toLowerCase() &&
        this.sortChars(candidate) === this.sortChars(this.subject)
      );
    });
  }

  sortChars(string) {
    let chars = string.toLowerCase().split("");
    chars.sort();
    return chars.join("");
  }
}

File: ruby/anagram/anagram.rb
class Anagram
  def initialize(input)
    @input = input.downcase
  end

  def match(candidates)
    candidates.select do |candidate|
      candidate.downcase != input &&
      candidate.downcase.chars.sort == input.chars.sort
    end
  end

  private

  attr_reader :input
end
